progiter/progiter.py

```diff
@@ -470,29 +485,74 @@ def _iterate(self):
         # Iterating is performance sensitive, so separate both cases - where
         # 'freq' is used and checks can be fast, and where 'adjust' is used and
         # checks need more calculation. This is worth duplicating code for.
-        if self.adjust:
-            for self._iter_idx, item in gen:
-                yield item
-
-                between_idx = (self._iter_idx - self._now_idx)
-                need_display = between_idx >= self.freq
 
-                # No clue how much time has passed, the frequency may be way off.
-                # If 'freq' is too large, checking time is necessary to notice it.
-                if not need_display:
-                    between_time = default_timer() - self._now_time
-                    need_display = between_time >= self.time_thresh
-
-                if need_display:
-                    # update progress information every so often
-                    self._update_and_display_message()
-        else:
+        use_fast_path = True
+        if self.adjust:
+            use_fast_path = False
+
+            homogeneous = self.homogeneous
+            try:
+                if homogeneous == 'auto':
+                    homogeneous = False
+                    # Take a few steps in the slow path and then check to see
+                    # if we should continue or do go down the fast path.
+                    num_initial_steps = 5
+
+                    # A call to time is 50ns, we can accept the overhead if it
+                    # is only .01% of the total loop time
+                    overhead_threshold = 50e-9 * 10_000
+
+                    _check_times = []
+                    for (self._iter_idx, item), _ in zip(gen, range(num_initial_steps)):
```

Comment:
Using `zip` causes the 5 elements to be generated and stored in memory, breaking the support for cases where there is only enough memory for one element at a time. `itertools.islice` has been in the standard library forever(?), so is probably ok to use.

Reply:
Good call.